sweep/sweeper.go

```diff
@@ -1,38 +1,690 @@
+
+	if result.Err == nil {
+		log.Debugf("Dispatching sweep success for %v to %v listeners",
+			outpoint, len(listeners))
+	} else {
+		log.Debugf("Dispatching sweep error for %v to %v listeners: %v",
+			outpoint, len(listeners), result.Err)
 	}
+
+	// Signal all listeners. Channel is buffered. Because we only send once
+	// on every channel, it should never block.
+	for _, resultChan := range listeners {
+		resultChan <- result
+	}
+
+	// Cancel spend notification with chain notifier. This is not necessary
+	// in case of a success, except for that a reorg could still happen.
+	if pendInput.ntfnRegCancel != nil {
+		log.Debugf("Canceling spend ntfn for %v", outpoint)
+		pendInput.ntfnRegCancel()
+	}
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
+
+// getInputLists goes through all pending inputs and constructs sweep lists,
+// each up to the configured maximum number of inputs. Negative yield inputs are
+// skipped. Transactions with an output below the dust limit are not published.
+// Those inputs remain pending and will be bundled with future inputs if
+// possible.
+func (s *UtxoSweeper) getInputLists(currentHeight int32,
+	satPerKW lnwallet.SatPerKWeight) ([]inputSet, error) {
+
+	// Filter for inputs that need to be swept. Create two lists: all
+	// sweepable inputs and a list containing only the new, never tried
+	// inputs.
+	//
+	// We want to create as large a tx as possible, so we return a final set
+	// list that starts with sets created from all inputs. However, there is
+	// a chance that those txes will not publish, because they already
+	// contain inputs that failed before. Therefore we also add sets
+	// consisting of only new inputs to the list, to make sure that new
+	// inputs are given a good, isolated chance of being published.
+	var newInputs, retryInputs []Input
+	for _, input := range s.pendingInputs {
+		// Skip inputs that have a minimum publish height that is not
+		// yet reached.
+		if input.minPublishHeight > currentHeight {
+			continue
+		}
+
+		// Add input to the either one of the lists.
+		if input.publishAttempts == 0 {
+			newInputs = append(newInputs, input.input)
+		} else {
+			retryInputs = append(retryInputs, input.input)
+		}
+	}
+
+	// If there is anything to retry, combine it with the new inputs and
+	// form input sets.
+	var allSets []inputSet
+	if len(retryInputs) > 0 {
+		var err error
+		allSets, err = generateInputPartitionings(
+			append(retryInputs, newInputs...),
+			s.relayFeePerKW, satPerKW,
+			s.cfg.MaxInputsPerTx,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create sets for just the new inputs.
+	newSets, err := generateInputPartitionings(
+		newInputs,
+		s.relayFeePerKW, satPerKW,
+		s.cfg.MaxInputsPerTx,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Append the new sets at the end of the list, because those tx likely
+	// have a higher fee per input.
+	return append(allSets, newSets...), nil
```

Comment:
I'm not really sure what is happening with these returned values. First the `allSets` will be attampted published, if that fails, then we'll publish `newSets`?

Reply:
Everything will be attempted always, but in that order. So if the allSets get accepted, we prefer that, because those include more inputs and likely result in a lower fee per input. 

The whole idea with the sweeper as it is now, is that it isn't a problem if publishtx fails. It spares us from predicting what is happening in the memppol.

Reply:
An improvement to this could be to sort the txes in order of fee per input.

Reply:
Won't we have duplicated inputs here since we use the `newInputs` slice twice to generate the set of candidate inputs? Or do we not care since if we double spend, then it'll fail and be retried up to the limit? Even with this design in mind, seems a bit wasteful to attempt a broadcast that we almost _know_ will fail since we already have a pending transaction that doubles spends our own inputs.

Reply:
It could be that some of the retry inputs have confirmed in the mean time and grouping them remaining retry inputs with the new inputs may lead to a single tx succesfully sweeping all. 

It depends on what the cost really is of a publish attempt. I assumed it is low.